### PR TITLE
ci: FIR-9440 automated release workflow

### PR DESCRIPTION
**Pull request by ptiurin**
Automating version increase and release to github and pypi.

Before merge we'll need to add pypi credentials to github secrets.
Token will be added in username/password secrets according to https://test.pypi.org/help/#apitoken

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        git tag ${{ steps.tag_generation.outputs.new_tag }}
+        git push origin ${{ steps.tag_generation.outputs.new_tag }}
+        
+    - name: Publish to pypi
+      run: |
+        python3 -m build
+        python3 -m twine upload dist/*
+      env:
+        TWINE_USERNAME: ${{ secrets.PYPI_USERNAME }}
+        TWINE_PASSWORD: ${{ secrets.PYPI_PASSWORD }}
```
Comment (eg-firebolt) [inline .github/workflows/release.yml:65]: new line at end of file?

File: ci/generate_version_tag.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        else:
+            # Enhancements bump minor version while anything else is considered a patch
+            enhancements = [c for c in changes.split("\n") if c.startswith("enh:")]
+            new_ver = ver.bump_minor() if enhancements else ver.bump_patch()
+    print(new_ver)
+
+
+def parse_arguments():
+    parser = argparse.ArgumentParser(
+        description="Determine the next tag based on change tags"
```
Comment (eg-firebolt) [inline ci/generate_version_tag.py:36]: Some additional context about this script would be helpful either here or in a docstring at the top of the file.

File: ci/generate_version_tag.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import argparse
+
+import semver
+
+
+def main(changes, old_tag, prerelease_tag, major_release):
```
Comment (eg-firebolt) [inline ci/generate_version_tag.py:6]: Could we add typehints and a usual docstring that describes each parameter?

Comment (ptiurin) [reply ci/generate_version_tag.py]: Done.

File: ci/generate_version_tag.py (hunk 2 above)
Comment (ptiurin) [reply ci/generate_version_tag.py]: Fair point, added a docstring.

File: ci/generate_version_tag.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    ver = semver.VersionInfo.parse(old_tag)
+    if major_release:
+        new_ver = ver.bump_major()
+    else:
+        if not changes:
+            raise Exception("No changes since the last release!")
+        # Breaking changes are of form <tag>!: <description>
+        # To avoid automatically publishing major version they need to be
+        # explicitly specified
+        breaking = [c for c in changes.split("\n") if "!:" in c]
```
Comment (stepansergeevitch) [inline ci/generate_version_tag.py:43]: We need to make sure that no one has `!:` in their PR name accidentally (i guess that's pretty unlikely)
Other solution could be to do `c.split(':', 1)[0].endswith('!')`

File: .github/workflows/integration-tests.yml [hunk 5]
```diff
@@ -1,6 +1,7 @@
 name: Run integration tests
 on:
   workflow_dispatch:
+  workflow_call:
```
Comment (stepansergeevitch) [inline .github/workflows/integration-tests.yml:4]: Are we reusing this workflow somewhere?

**Review by stepansergeevitch — APPROVED** [review]

File: .github/workflows/integration-tests.yml (hunk 5 above)
Comment (ptiurin) [reply .github/workflows/integration-tests.yml:4]: Good spot, I was meant to add integration test check before release, but removed it for testing.

File: ci/generate_version_tag.py (hunk 4 above)
Comment (ptiurin) [reply ci/generate_version_tag.py:43]: Yeah, doesn't look like it's spotted by the PR title checker so better be safe.

**Pull request merged**